providers/hedns/hednsProvider.go

```diff
@@ -0,0 +1,622 @@
+	}
+
+	// Ensure the order is deterministic
+	sort.Strings(domains)
+
+	return domains, err
+}
+
+func (c *ApiClient) EnsureDomainExists(domain string) error {
+	domains, err := c.ListZones()
+	if err != nil {
+		return err
+	}
+
+	for _, d := range domains {
+		if d == domain {
+			return nil
+		}
+	}
+
+	return c.createDomain(domain)
+}
+
+func (c *ApiClient) GetNameservers(_ string) ([]*models.Nameserver, error) {
+	return models.ToNameservers(defaultNameservers)
+}
+
+func (c *ApiClient) GetDomainCorrections(dc *models.DomainConfig) ([]*models.Correction, error) {
+	var corrections []*models.Correction
+
+	err := dc.Punycode()
+	if err != nil {
+		return nil, err
+	}
+
+	records, err := c.GetZoneRecords(dc.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the SOA record to get the ZoneId, then remove it from the list.
+	zoneId := uint64(0)
+	var prunedRecords models.Records
+	for _, r := range records {
+		if r.Type == "SOA" {
+			zoneId = r.Original.(Record).ZoneId
+		} else {
+			prunedRecords = append(prunedRecords, r)
+		}
+	}
+
+	// Normalize
+	models.PostProcessRecords(prunedRecords)
+
+	differ := diff.New(dc)
+	_, toCreate, toDelete, toModify, err := differ.IncrementalDiff(prunedRecords)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, del := range toDelete {
+		record := del.Existing
+		corrections = append(corrections, &models.Correction{
+			Msg: del.String(),
+			F:   func() error { return c.deleteZoneRecord(record) },
+		})
+	}
+
+	for _, cre := range toCreate {
+		record := cre.Desired
+		record.Original = Record{
+			ZoneName:   dc.Name,
+			ZoneId:     zoneId,
+			RecordName: cre.Desired.Name,
+		}
+		corrections = append(corrections, &models.Correction{
+			Msg: cre.String(),
+			F:   func() error { return c.editZoneRecord(record, true) },
+		})
+	}
+
+	for _, mod := range toModify {
+		record := mod.Desired
+		record.Original = Record{
+			ZoneName:   dc.Name,
+			ZoneId:     zoneId,
+			RecordId:   mod.Existing.Original.(Record).RecordId,
+			RecordName: mod.Desired.Name,
+		}
+		corrections = append(corrections, &models.Correction{
+			Msg: mod.String(),
+			F:   func() error { return c.editZoneRecord(record, false) },
+		})
+	}
+
+	return corrections, err
+}
+
+func (c *ApiClient) GetZoneRecords(domain string) (models.Records, error) {
+	var zoneRecords []*models.RecordConfig
+
+	// Get Domain ID
+	domains, err := c.listDomains()
+	if err != nil {
+		return nil, err
+	}
+
+	domainId, domainExists := domains[domain]
+	if !domainExists {
+		return nil, fmt.Errorf("domain %s does not exist", domain)
+	}
+
+	queryUrl, _ := url.Parse(ApiUrl)
+	q := queryUrl.Query()
+	q.Add("hosted_dns_zoneid", strconv.FormatUint(domainId, 10))
+	q.Add("menu", "edit_zone")
+	q.Add("hosted_dns_editzone", "")
+	queryUrl.RawQuery = q.Encode()
+
+	response, err := c.httpClient.Get(queryUrl.String())
+	if err != nil {
+		return nil, err
+	}
+	defer response.Body.Close()
+
+	document, err := goquery.NewDocumentFromReader(response.Body)
```

Comment:
Wow, they don't offer json or anything?  That's pretty sadistic to require us to parse HTML.   I should add that to my [API Practices If You Hate Your Customers](https://queue.acm.org/detail.cfm?id=3375635) article.

Anyway... if that's all the offer, I salute your use of goquery to do the heavy lifting!

Reply:
Oh wait... are we parsing a web UI meant for humans?  If so, will this break if they change their web UI?  (its ok if the answer is yes, but there should be a note about that somewhere in the .md doc.)

Reply:
Yes unfortunately that is the case, I should have made that clear from the outset. I have added an "Important Note" at the top of `hedns.md` indicating this. It might break if their interface changes, but it's not changed in a long-time save to add new record types and I've added some sanity error checks which fail quickly if they do. The selectors are also relative loose so they could hopefully change a fair bit without them failing.

Ultimately, if this is a reason to not accept this PR I'd understand :), I mostly wrote this for my own use, as I'm fed up of using the web interface and figured it would be useful for others.

Reply:
They have such a great reputation for technical excellence.  I'm kind of shocked.

That said... yes, I'll accept it but if anything breaks, you'll be assigned the bug :-)

Ready for review/merge?

Reply:
You're not the only one who's shocked :) Hopefully, this will convince them a proper API would be worthwhile, the calls have gone unanswered so far!

Yes, I'm happy for review/merge.